### ISSUE TITLE
Question: Benchmarking UI performance

### ISSUE THREAD

**Issue by chingucoding** [microsoft/microsoft-ui-xaml#2276]
Currently there seems to be no good way to create a reliable benchmark of UI's. You can use the Performance profiler of Visual Studio, however to get a reliable result you need to rerun it multiple times to eliminate the factor of other apps on the PC stealing CPU usage.

Are there any way to do this? A good benchmark would include the time spent on different layout stages and also list the memory usage of individual Visual Tree items.

For .Net there is BenchmarkDotNet which is a very good tool for benchmarking, however they don't support UWP apps. Would this be something that changes when we use .Net 5?

@stevenbrix FYI

**Comment by Austin-Lamb** [comment]
@chingucoding - could you describe more about what you'd like to profile?  Memory usage?  CPU cycles?  Disk I/O?

You mentioned wanting to see time spent during "different layout stages" - can you clarify what you mean there?  Do you mean just measure vs. arrange?  Or would you want to see template expansion as a separate 'stage'?  What about phased rendering (x:Phase)?

We are working on a perf lab to measure WinUI itself but not apps - it's a good question about BenhcmarkDotNet, I'll ask some folks internally to see what the plans may be there.

Your suggestion about seeing the memory usage of a subtree is a good one - I'll work on spinning that idea out into a separate feature request so we can track it better (and I'll post a link back here once I do).  Thanks for that!

**Comment by stevenbrix** [comment]
Adding @billwert from the .NET perf team in case he has any ideas

**Comment by chingucoding** [comment]
The main things I would like to profile are Memory, CPU usage/cycles, I/O and networking.

Regarding the different layout stages: Yes I would mainly think about measure and arrange, since there aren't any other big UI lifecycle stages right? Phasing could indeed be interesting, but I think it would be to much right now, and would consider it more of a nice to have. 

Something that might also be interesting is to see how much time is spent interacting with the VisualTree, e.g. finding childtemplates and how much time is being spent inside VisualStateManagers. But I would consider those a nice to have too.

Looking forward to the visual tree memory usage feature request/proposal!

**Comment by stevenbrix** [comment]
All of this sounds like information we gather in the Diagnostics Tools window, but I'm not sure how discoverable it is.

Adding @SvetBonev and @srdjanjovcic who are the owners of those tools.

**Comment by billwert** [comment]
Hello!

As you point out the Diagnostics window has lots of good information here. It works by parsing the ETW events emitted by the UWP platform to show the breakdowns in question.

There's not a good off the shelf solution to this, sadly. Benchmark.NET is not really what you want for this, though it can be used for the purpose; it's much better at microbenchmark sort of things. 
I don't know the current state of the VS performance command line tools. VSPerfCmd doesn't immediately have anything obvious to help here.

In the .NET team's analysis we tend to use [PerfView](https://github.com/microsoft/perfview) for lots of this, but it isn't as nice as the VS experience. There's also a component in that repo called TraceEvent that makes it easy to write small custom parsers. If I was trying to benchmark my UI app in a repeated fashion I might instrument the parts I cared most about using an [`EventSource`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventsource?view=netcore-3.1) to mark the parts I cared most about and then write some tooling to do repeated runs and parse the resultant trace. This is how our startup testing works, [for example](https://github.com/dotnet/performance/blob/master/src/tools/ScenarioMeasurement/Startup/GenericStartupParser.cs).

The VS tools folks who've been added can hopefully help more.
